### ISSUE TITLE
[MLIR][Frontend] Differentiating `qml.QubitUnitary` results in core dump.

### ISSUE THREAD

**Issue by erick-xanadu** [PennyLaneAI/catalyst#384]
* *Expected behavior:* Attempting to differentiate a non-differentiable gate (like `qml.QubitUnitary`) should ideally raise an error instead of a core dump.

* *Actual behavior:* A segfault during the enzyme passes.

#### Source code and tracebacks

```python
import pennylane as qml
from catalyst import measure, qjit, debug, grad
from catalyst.pennylane_extensions import qfunc
from jax import numpy as jnp
from jax.core import ShapedArray as tensor
import jax
from catalyst.utils.patching import Patcher

def _decomp_to_unitary(self, *_args, **_kwargs):
    return [qml.QubitUnitary(qml.matrix(self), wires=self.wires)]

with Patcher(
        (qml.SingleExcitation, "decomposition", _decomp_to_unitary),
):

    @qml.qnode(qml.device("lightning.qubit", wires=3))
    def f(param0):
        qml.SingleExcitation(param0, wires=[0, 1])
        return qml.expval(qml.PauliZ(2))

    @qjit
    def foo(param0):
        diff = grad(f, argnum=0)
        h = diff(param0)
        return h[0]

    print(foo(1.0))
```

```
// LLVM Dump followed by:


python: /home/erick.ochoalopez/Code/cataliist/mlir/Enzyme/enzyme/Enzyme/EnzymeLogic.cpp:2262: const AugmentedReturn &EnzymeLogic::CreateAugmentedPrimal(llvm::Function *, DIFFE_TYPE, ArrayRef<DIFFE_TYPE>, TypeAnalysis &, bool, bool, const FnTypeInfo &, const std::vector<bool>, bool, unsigned int, bool, bool): Assertion `0 && "attempting to differentiate function without definition"' failed.
Aborted (core dumped)
```

**Comment by dime10** [comment]
This is known to be unsupported, but we should raise a Python error that the function cannot be differentiated.
In general I was thinking it would be a good idea to add a program verifier in the frontend to make sure the compiler and device capabilities are respected by the user program.

**Comment by dime10** [comment]
Just looked into this and there is a regression introduced in #360 where the quantum register we dealloc is now its last SSA value rather than the first SSA value. In certain scenarios, like here in the presence of `QubitUnitary`, our gradient transform has a harder time eliminating quantum instructions. This results in quantum instruction calls being present in the pre-processing function that Enzyme is differentiating. These instructions don't have a definition, hence the resulting error `Assertion '0 && "attempting to differentiate function without definition"' failed.`

**Comment by josh146** [comment]
Something to note is that `QubitUnitary` is 'differentiable' via the parameter-shift rule if it is decomposed down into parameter-shift compatible gates (this is what PL Python does iirc).

In addition, it's also differentiable via finite-difference, but it take additional work to code this up (e.g., allowing finite-diff to work well on array elements), and I would say is not worth it/valuable -- we never supported this in PL.

**Comment by erick-xanadu** [comment]
@dime10 I re-ran the example and there was no failure, Do you think we can close this ticket?

**Comment by dime10** [comment]
It's true that the input no longer crashes the compiler, but it would still produce incorrect results. What I'd like to add is a program checker that will error if the user program is incompatible (or could not be made compatible) for the chosen differentiation method. I think that would be a better resolution to this ticket (and I notice that is also what you put in _expected behaviour_).

**Comment by dime10** [comment]
The verification feature has been implemented :)
